[WIP] Decent offline functionality and other improvements

di72nn opened the pull request:
Ok, I finally started to implement it.

The first batch of changes:
* Update and add some dependencies: update greendao, okhttp, add greenrobot:eventbus.
* Update DB schema (add QueueItem), regenerate greendao classes.
* Add an IntentService and some basic logic to perform network-related operations in BG service and to enqueue failed actions for later sync.

I create this PR so we can discuss something about it in the future.
Currently I have some thing to implement, so code-related collaboration is not yet needed (not that I expect much).

TODO:
- [x] Use background service for the most operations.
- [x] Handle offline operations.
- [x] Basic auto-sync.
- [x] Decent notifications.
- [x] Basic PreferenceFragment-based settings.
- [x] Set `configurationIsOk` flag somewhere (requires configuration master/wizard?).
- [x] Unset `configurationIsOk` flag in Settings.
- [x] Delete obsolete translation lines.

Not so urgent/important TODO:
 - [Better temporary errors detection](#issuecomment-230098478).
 - Better/prettier/fancier notifications.
 - Do something with UI responsiveness (while background operation is performed).
 - Refactoring: reuse code.

Separate features/improvements:
 - Show last update time somewhere.
 - Differentiate Wi-Fi and data connection in auto-update?
 - Download PDF in BGService.
 - Use support library for Settings?

tcitworld commented on gen-dao/src/main/java/fr/gaulupeau/apps/Poche/GenerateDao.java:
```diff
@@ -17,14 +17,23 @@ public static void main(String[] args) throws Exception {
-        article.addBooleanProperty("sync").columnName("sync");
-        article.addDateProperty("updateDate").columnName("update_date");
+        article.addBooleanProperty("sync").columnName("sync"); // TODO: remove
+        article.addDateProperty("updateDate").columnName("update_date"); // TODO: check: what is this prop for?
```
I think it was to display somehow the date the article had been added through v1's RSS feeds. Not relevant until we use the API from v2.

di72nn commented:
Another update:

Implement and replace the following actions:
 * archive/unarchive article,
 * favorite/unfavorite article,
 * delete article,
 * add link.

Implement queue sync (client->server) (triggered via menu item).

A lot of similar code - I didn't get to code reuse yet.

Network-error handling is very basic (false positives on error detection are possible - some items may stuck in offline queue) and no notifications (only debug toasts), but I think it already may be tested.

tcitworld commented on app/src/main/java/fr/gaulupeau/apps/Poche/service/EventProcessor.java:
```diff
@@ -0,0 +1,283 @@
+        ActionResult result = event.getResult();
+
+        // TODO: fix: debug only implementation
+
```
Maybe a notification & a toast each time you do something as basic as archiving/favoriting is too much ?

di72nn commented:
> Maybe a notification & a toast each time you do something as basic as archiving/favoriting is too much ?

That is totally a debug-only stuff. I wanted to see when something happened and that there is no problems with notifications.

tcitworld commented:
Great work ! I didn't run into major issues while trying, and I'll try to debug a bit later this week.

Can you also put an `AlarmManager` in order to make background sync run each time in a specified amount of time ? That would be quite nice.

di72nn commented:
Yep, I plan to improve background sync.

BTW, do you experience [relatively] long transition from ArticleReadActivity after "Mark as Read" was pressed? On my device with Android 4.4 the background operation seems to have priority over UI update somehow. I didn't look into it yet, since things still may change - I'm just curious yet.

tcitworld commented:
Nope, it changes instantly. But I have a good Internet connection, so I can't be too sure.

di72nn commented:
That is more noticeable when "Mark as Read" is the first operation since app launch (HttpClient was not yet authorized, so it makes more requests). But, I guess, it won't be much slower. Good to know.

tcitworld commented:
Nope, no difference there.

di72nn commented:
Testing needed:
I have a class of errors named "Temporary", which should include errors like connection break, random connection timeout and such (all exceptions are probably subclasses of IOException). As of now I mark IOException (and the subclasses) as "Unknown" error type, so it will be shown in a notification. I need to gather a list of various errors that shown as "Unknown" but should be considered as "Temporary".

I also removed some unnecessary notifications to improve usage of dev version.

di72nn commented:
I guess it's also a time to discuss some potential UX changes.
I think we should have a "configuration is correct" flag. The flag is supposed to indicate that the saved settings allow the app to connect to the server -- at least a server URL and credentials are correct.
It may be set if a user saves settings after successful connection test or the user may be forced to go through a connection wizard in order to use the app (requires some more work on UI). There is also an option to set this flag right after the first successful network request, but I don't like this idea for a bunch or reasons.

Having this flag gives the following advantages:
* Allows to Remove weird checks like `url != "https://"`, `username != null`, and so on.
* Allows to make decisions in cases like `UnknownHostException` (we can be pretty sure that it happens because of temporary DNS-not-yet-available error; if the host gets removed from DNS, it's not the apps problem really, and it's not the app's responsibility to report it).
* It may be used to handle other errors better (I have no examples yet).
Cons:
* We need to somehow force the user to check connection **once**.
I can't remember any more cons right now.

One clarification: if the user changes his password on the server, it can be detected pretty reliably and shown anyway.

tcitworld commented:
> It may be set if a user saves settings after successful connection test or the user may be forced to go through a connection wizard in order to use the app (requires some more work on UI).

Yup. That's one of the two things I want to make in the next weeks. This and #143.

Totally agree with you on everything else.

tcitworld commented:
Tested a few things : 

* Needs a toast message when an entry is added to the queue when shared from external app.
* When an article is favorited from the web interface, and feeds updated inside the app. [The article](http://www.liberation.fr/debats/2016/06/17/ce-gouvernement-a-peur_1460153) in question contains a `<script>` tag that throws an exception that was handled before.
```
07-02 15:46:43.461 10118-17270/fr.gaulupeau.apps.InThePoche.debug V/FeedUpdater: processFeed() parser.getName()=item
07-02 15:46:43.462 10118-17270/fr.gaulupeau.apps.InThePoche.debug E/FeedUpdater.updateByF: XmlPullParserException on Favorite
                                                                                           org.xmlpull.v1.XmlPullParserException: expected: /description read: script (position:END_TAG </script>@310:582 in java.io.InputStreamReader@8c1bb35) 
                                                                                               at org.kxml2.io.KXmlParser.readEndTag(KXmlParser.java:997)
                                                                                               at org.kxml2.io.KXmlParser.next(KXmlParser.java:383)
                                                                                               at org.kxml2.io.KXmlParser.next(KXmlParser.java:321)
                                                                                               at org.kxml2.io.KXmlParser.nextText(KXmlParser.java:2138)
                                                                                               at fr.gaulupeau.apps.Poche.network.FeedUpdater.parseNextText(FeedUpdater.java:405)
                                                                                               at fr.gaulupeau.apps.Poche.network.FeedUpdater.parseItem(FeedUpdater.java:391)
                                                                                               at fr.gaulupeau.apps.Poche.network.FeedUpdater.processFeed(FeedUpdater.java:270)
                                                                                               at fr.gaulupeau.apps.Poche.network.FeedUpdater.updateByFeed(FeedUpdater.java:176)
                                                                                               at fr.gaulupeau.apps.Poche.network.FeedUpdater.updateInternal(FeedUpdater.java:146)
                                                                                               at fr.gaulupeau.apps.Poche.network.FeedUpdater.update(FeedUpdater.java:88)
                                                                                               at fr.gaulupeau.apps.Poche.service.BGService.updateFeed(BGService.java:525)
                                                                                               at fr.gaulupeau.apps.Poche.service.BGService.onHandleIntent(BGService.java:91)
                                                                                               at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:66)
                                                                                               at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                               at android.os.Looper.loop(Looper.java:148)
                                                                                               at android.os.HandlerThread.run(HandlerThread.java:61)
07-02 15:46:43.464 10118-17270/fr.gaulupeau.apps.InThePoche.debug W/BGService: updateFeed() FeedUpdaterException
                                                                               fr.gaulupeau.apps.Poche.network.FeedUpdater$FeedUpdaterException: Erreur de traitement du flux
                                                                                   at fr.gaulupeau.apps.Poche.network.FeedUpdater.updateByFeed(FeedUpdater.java:183)
                                                                                   at fr.gaulupeau.apps.Poche.network.FeedUpdater.updateInternal(FeedUpdater.java:146)
                                                                                   at fr.gaulupeau.apps.Poche.network.FeedUpdater.update(FeedUpdater.java:88)
                                                                                   at fr.gaulupeau.apps.Poche.service.BGService.updateFeed(BGService.java:525)
                                                                                   at fr.gaulupeau.apps.Poche.service.BGService.onHandleIntent(BGService.java:91)
                                                                                   at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:66)
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                                   at android.os.Looper.loop(Looper.java:148)
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61)
07-02 15:46:43.468 10118-17270/fr.gaulupeau.apps.InThePoche.debug D/EventProcessor: onUpdateFeedsFinishedEvent() started
07-02 15:46:43.469 10118-17270/fr.gaulupeau.apps.InThePoche.debug D/BGService: updateFeed() finished
```

di72nn commented:
> Needs a toast message when an entry is added to the queue when shared from external app.
Added to my todo list.

Regarding the error: it seems to be server side issue. I can't reproduce it, according to history I didn't change anything in XML parsing, and all the `<description>` content expected to be `<![CDATA[`-enclosed (v2 server) or tags escaped (v1 server).

di72nn commented:
Added basic auto sync.
Should be noted: in this dev version "is configuration ok" flag is set on "Save" button click in Settings after "Connection test" was successful (if connection test was not performed, flag is not updated; if connection test failed, flag will be set to false).

tcitworld commented on app/src/main/java/fr/gaulupeau/apps/Poche/events/EventProcessor.java:
```diff
@@ -49,6 +51,45 @@ public void stop() {
+
+        if(!WallabagConnection.isNetworkOnline()) {
+            Log.d(TAG, "No network, skipping auto-sync");
+            // TODO: set another closer alarm?
```
I don't find it necessary. If I understand correctly, `networkChanged` will sync things anyway as soon as the network is on ?

tcitworld commented on app/src/main/java/fr/gaulupeau/apps/Poche/service/AlarmHelper.java:
```diff
@@ -0,0 +1,60 @@
+public class AlarmHelper {
+
+    public static void setAlarm(Context context, long interval, boolean enableReceiver) {
+        // TODO: maybe use last update time to calculate initial delay
```
:+1: 
Yup, it would be nice to show last update time.

tcitworld commented:
>  Should be noted: in this dev version "is configuration ok" flag is set on "Save" button click in Settings after "Connection test" was successful (if connection test was not performed, flag is not updated; if connection test failed, flag will be set to false).

I think the "Save" button could be disabled while `configurationIsOk != success` ?

di72nn commented:
> I don't find it necessary. If I understand correctly, networkChanged will sync things anyway as soon as the network is on ?

I disabled connectivity-changed receiver for now. I was thinking about various auto-sync options, but I wasn't very productive to implement all of it in one sitting (I also wasn't sure what behavior is desired), so I started with basic auto-sync based on AlarmManager only.

> I think the "Save" button could be disabled while configurationIsOk != success ?

Disabling the "Save" button in the only settings view is not a good idea. If it was a separate view to set up URL and credentials it might be viable, but not in our situation.
Anyway, current behavior is intended for development purposes - so that it is possible to easily set this flag to false.

tcitworld commented:
OK.
Autosync works fine for me. :-)

di72nn commented:
It probably should be tested, but I added it in case if auto-sync is triggered right after device wake up - when network may not be available yet. I added a debug message to check it.

di72nn commented:
I did not delete string resources and translations just in case we decide to use some of them before this PR comes to it's end.

tcitworld commented:
Okay. I edited first message to add a reminder.

Strubbl commented:
> BTW, do you experience [relatively] long transition from ArticleReadActivity after "Mark as Read" was pressed?

Yes. On Sony Xperia Z5c. UI is blocked until the operation has finished. So the operation runs in main/UI thread?


Also if i successfully sync, the widget gets no update. And the widget gets not update when i mark articles as read.

di72nn commented:
> UI is blocked until the operation has finished. So the operation runs in main/UI thread?

It is highly unlikely. More like an IntentService thread is prioritized over the main thread. At some point I even "fixed" it by setting lower priority of the IntentService's thread, but it somehow stopped working after some related changes. It seems I'll have to investigate the problem after all.

> the widget gets no update

I presume `onFeedsChangedEvent() started; updating widget` is not present in logs? I don't remember whether I actually tested the widget :/

Strubbl commented:
> I presume onFeedsChangedEvent() started; updating widget is not present in logs?

Sorry, i did not test it yet with logcat attached.

di72nn commented:
The last commit message is a bit misleading: I only made some under the hood changes needed to implement PreferenceActivity/PreferenceFragment-based settings.

tcitworld commented:
That's great ! We can finally migrate to Preference Fragments ! Many thanks for all your work on this PR.

di72nn commented:
Added basic PreferenceFragment-based settings (alongside the old Settings). I had a hard time implementing it with `com.android.support:preference-v7`, so I decided to start with the default PreferenceFragment. We may want to use some kind of a third-party library for that (like these: [1](https://github.com/codevscolor/MaterialPreference), [2](https://github.com/consp1racy/android-support-preference), [3](https://github.com/Gericop/Android-Support-Preference-V7-Fix)).

The settings UI is pretty basic (lacks most of the summaries and such) -- it is easy to add so I don't like to do such work. Help is welcome.

Oh, and I didn't test it thoroughly.

di72nn commented:
Added a Connection Wizard to simplify initial setup and to avoid some of popular configuration errors. Suggestions regarding UI and UX are welcome.

tcitworld commented:
\o/ Checking out and building...

tcitworld commented:
That's great ! Here's a few returns : 

* Please add v2.wallabag.org to the preconfigured providers
* Wiping the database and relaunching the app doesn't trigger the Connection Wizard
* The « Feeds Configuration » steop is not of any use.
* When the wizard is finished, don't ask anymore if we want to sync since the list of articles is empty, do it automatically


Also, got an issue while syncing feeds (restarting the app showed the articles, though) : 
```
09-07 11:12:47.792 3084-4367/fr.gaulupeau.apps.InThePoche.debug E/BGService: updateFeed() XmlPullParserException
org.xmlpull.v1.XmlPullParserException: expected: /description read: script (position:END_TAG </script>@310:582 in java.io.InputStreamReader@b562ec) 
at org.kxml2.io.KXmlParser.readEndTag(KXmlParser.java:984)
at org.kxml2.io.KXmlParser.next(KXmlParser.java:375)
at org.kxml2.io.KXmlParser.next(KXmlParser.java:313)
at fr.gaulupeau.apps.Poche.network.FeedUpdater.skipElement(FeedUpdater.java:339)
at fr.gaulupeau.apps.Poche.network.FeedUpdater.processFeed(FeedUpdater.java:316)
at fr.gaulupeau.apps.Poche.network.FeedUpdater.updateByFeed(FeedUpdater.java:169)
at fr.gaulupeau.apps.Poche.network.FeedUpdater.updateAllFeeds(FeedUpdater.java:114)
at fr.gaulupeau.apps.Poche.network.FeedUpdater.update(FeedUpdater.java:82)
at fr.gaulupeau.apps.Poche.service.BGService.updateFeed(BGService.java:540)
at fr.gaulupeau.apps.Poche.service.BGService.onHandleIntent(BGService.java:113)
at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67)
at android.os.Handler.dispatchMessage(Handler.java:102)
at android.os.Looper.loop(Looper.java:154)
at android.os.HandlerThread.run(HandlerThread.java:61)
```

Here's a few screenshots for those wondering.
![screenshot_20160907-130934](https://cloud.githubusercontent.com/assets/2197836/18310051/cb8fd534-74fc-11e6-983f-cfee83847ac1.png)
![screenshot_20160907-130940](https://cloud.githubusercontent.com/assets/2197836/18310053/cb924cce-74fc-11e6-9d17-eea301992c8e.png)
![screenshot_20160907-131002](https://cloud.githubusercontent.com/assets/2197836/18310052/cb90a2d4-74fc-11e6-99fc-4f96d50131d1.png)
![screenshot_20160907-131012](https://cloud.githubusercontent.com/assets/2197836/18310054/cb92d5b8-74fc-11e6-81e2-b4ebcb0033ca.png)
![screenshot_20160907-131015](https://cloud.githubusercontent.com/assets/2197836/18310050/cb8f0708-74fc-11e6-9f96-0521394eb376.png)

Strubbl commented:
Looks nice so far.

I have one issue: The widget gets no update on updating of unread feed finished. 

It would be nice if we can merge this as soon as major issues are fixed, to base new work on this commits. Also gradle and libs were updated already in here. It would be nice to see that on master.

di72nn commented:
@tcitworld 
> Please add v2.wallabag.org to the preconfigured providers

Done.

> The « Feeds Configuration » step is not of any use.

Merged with the previous one.

> Wiping the database and relaunching the app doesn't trigger the Connection Wizard
> When the wizard is finished, don't ask anymore if we want to sync since the list of articles is empty, do it automatically

Noted, I'll check it later.

> got an issue while syncing feeds

Check [my previous answer](#issuecomment-230105731).

@Strubbl 
Right, I'll try to focus on the most urgent problems.

di72nn commented:
* Fixed widget updating.
* Updated libraries and Android SDK level to 24.
* CI build fails because of Android SDK level 24 ._.
* "UI freezes" should be reduced.
* Minor updates for New Settings. I consider Settings "decent" now.
* Added DB OpenHelper which performs DB migration (currently it only tries to migrate Offline URLs from the old version; **testing needed**).
* **Experimental** persistent cookies: added a small library to store cookies in `SharedPreferences`, enabled "remember me" for auth.

I'll check my notes for minor todo's and there is a cleanup to do, but other than that the PR seems to be more or less complete.

Strubbl commented:
I would suggest to leave out the "save settings" screen in case everything just goes fine (login & connection test etc.). I see no reason why the user should push the button to save settings or not. I ask, in which case he does not want save the settings and abort? What do you think?

Strubbl commented:
Widget updates are working fine again. Thanks a lot! :) I have one small issue with the widget counter of unread articles. My wallabag instance show i have 710 items unread, but the widget only shows 709. Even after wiping the app. Can you please check if it is the same in your setup? (Maybe it is only my setup which is wrong)

Strubbl commented:
Okay, now the reason for the failing travis are lint errors.

di72nn commented:
Yeah, I'm not sure about the last connection wizard screen either. As I see it, we should let the user know that everything completed well, rather than just throw him to some unrelated activity (the main screen). Though, I don't know what to put on that last screen exactly. Initially I planned to show a summary of selected/detected settings (I even called the screen "summary" internally), but then I realized that average user wouldn't be interested in the details.
I don't have a solid reason to have give the user an ability to abort connection wizard on the last step (well, someone may want to test new/alternative server setup, but it's a far-fetched guess).

Thanks to @Strubbl the CI build issue is resolved.
Though, now CI shows issues I don't get locally. On my local machine I get `Ran lint on variant debug: 333 issues found` and no errors, CI shows `Ran lint on variant debug: 334 issues found`.

Strubbl commented:
I think if we the configuration succeeded and we redirect the user directly to the main page and he sees the animation for downloading the feeds, that should be sufficient. Maybe we can write a toast after successful configuration (e.g. "Setup complete. Downloading articles") instead of that summary/save settings screen.


I can even top that CI lints. I checked out your latest commit here and ran lint from commandline :
```
$ ./gradlew :app:lint
...
:app:lint
Ran lint on variant debug: 336 issues found
Ran lint on variant release: 332 issues found
...
```

di72nn commented:
Thanks to @Strubbl the second CI issue is resolved.

I disabled the last Connection Wizard screen, but it can be enabled for some specific cases in the future (for some provider present, for example).
I also cleaned strings.xml.

Aside from stubs like `<string name="connectionWizard_welcome_text">Something welcoming here. Blah-blah-blah.</string>`, the PR seems to be ready for review.

Strubbl commented:
Why did you  Revert "travis: enable lint report as text on commandline"? Doesn't it make sense to leave that in to investigate future Travis CI issues?

di72nn commented:
We can always reapply it when needed. So far it was needed only once (as far as I know).

tcitworld commented:
Lint report error fixing is definitely something we can do, but indeed Travis doesn't needs it right now.

Strubbl commented:
I have review some code of this PR sporadicly. This PR is very big and i think i am not going to review every change in the 100+ files. From my point of view we can merge this and then i am going to have a closer look at the Google Developer console of the app store. 

I have tested the recent version of this PR in the last couple of days already. It works fine for me.

tcitworld commented:
Here we go. Will release on Monday (I'm not sure to have the keys on me today).
